Feature/issue 262 cvodes 2

wds15 opened the pull request:
#### Submisison Checklist

- [X] Run unit tests: `./runTests.py test/unit`
- [X] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary:
Add CVODES integrator

#### Intended Effect:
Provide faster ODE integrator which is also more robust.

#### How to Verify:
Tests for CVODE have been ported to CVODES. So look at the code and run the ODE related tests.

#### Side Effects:

#### Documentation:
doxygen comments are written, may need more.

#### Reviewer Suggestions: 
@bob-carpenter @betanalpha @syclik 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is requiring  to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

wds15 commented:
A question: The initial_state member function of cvodes_integrator is not needed by my code. It is only there to make tests happy which had been written for cvode - keep or remove? The initial state is created in the constructor of cvodes_integrator.

syclik commented:
Remove. Unless there's a good reason to expose it. 

> On Apr 17, 2016, at 4:13 AM, wds15 <notifications@github.com> wrote:
> 
> A question: The initial_state member function of cvodes_integrator is not needed by my code. It is only there to make tests happy which had been written for cvode - keep or remove? The initial state is created in the constructor of cvodes_integrator.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

wds15 commented:
Hmm.. Jenkins is not happy, but I think this is due to a Jenkins error when I read the log right.

I will remove the initial_state member function and I think there is a way to meaningfully replace the tests where it is involved.

wds15 commented:
Not sure here what upsets Jenkins...some issue with cmdstan or a Jenkins bug as far as I can tell

syclik commented on stan/math/rev/mat/functor/integrate_ode_cvodes.hpp:
```diff
@@ -0,0 +1,123 @@
+#include <stan/math/prim/arr/err/check_ordered.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/rev/mat/functor/cvodes_integrator.hpp>
+#include <stan/math/rev/arr/fun/decouple_ode_states.hpp>
```
missing an include for `stan/math/rev/mat/functor/ode_model.hpp`.

syclik commented on stan/math/rev/mat/functor/integrate_ode_cvodes.hpp:
```diff
@@ -0,0 +1,123 @@
+                         double rel_tol = 1e-10,
+                         double abs_tol = 1e-10,
+                         long int max_num_steps = 1e8,  // NOLINT(runtime/int)
+                         const size_t solver = 1,
```
@bob-carpenter, would you suggest using a enum for this?

bob-carpenter replied:
I'd actually rather see this broken out as separate functions
with shared function calls if there's work that's the same in each
one.  That way we don't get the 

```
case solver:
  ...

do something else

case solver:
  ...

...

pattern that's so hard to maintain and understand due to different
branches being linked.

> On Apr 20, 2016, at 10:13 AM, Daniel Lee <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/integrate_ode_cvodes.hpp:
> 
> > +     * @return a vector of states, each state being a vector of the
> > +     * same size as the state variable, corresponding to a time in ts.
> > +     */
> > +    template <typename F, typename T1, typename T2>
> > +    std::vector<std::vector<typename stan::return_type<T1, T2>::type> >
> > +    integrate_ode_cvodes(const F& f,
> > +                         const std::vector<T1>& y0,
> > +                         const double t0,
> > +                         const std::vector<double>& ts,
> > +                         const std::vector<T2> theta,
> > +                         const std::vector<double>& x,
> > +                         const std::vector<int>& x_int,
> > +                         double rel_tol = 1e-10,
> > +                         double abs_tol = 1e-10,
> > +                         long int max_num_steps = 1e8,  // NOLINT(runtime/int)
> > +                         const size_t solver = 1,
> 
> @bob-carpenter, would you suggest using a enum for this?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

wds15 replied:
So is only the solver argument to integrate_ode_cvodes a problem OR how it's done in cvodes_integrator? In any case, I would suggest for the moment to restrict things to the BDF solver only and we clean up the different solver options later.

bob-carpenter replied:
That sounds good.

> On Apr 20, 2016, at 5:15 PM, wds15 <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/integrate_ode_cvodes.hpp:
> 
> > +     * @return a vector of states, each state being a vector of the
> > +     * same size as the state variable, corresponding to a time in ts.
> > +     */
> > +    template <typename F, typename T1, typename T2>
> > +    std::vector<std::vector<typename stan::return_type<T1, T2>::type> >
> > +    integrate_ode_cvodes(const F& f,
> > +                         const std::vector<T1>& y0,
> > +                         const double t0,
> > +                         const std::vector<double>& ts,
> > +                         const std::vector<T2> theta,
> > +                         const std::vector<double>& x,
> > +                         const std::vector<int>& x_int,
> > +                         double rel_tol = 1e-10,
> > +                         double abs_tol = 1e-10,
> > +                         long int max_num_steps = 1e8,  // NOLINT(runtime/int)
> > +                         const size_t solver = 1,
> 
> So is only the solver argument to integrate_ode_cvodes a problem OR how it's done in cvodes_integrator? In any case, I would suggest for the moment to restrict things to the BDF solver only and we clean up the different solver options later.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

syclik commented on stan/math/rev/mat/functor/integrate_ode_cvodes.hpp:
```diff
@@ -0,0 +1,123 @@
+                                      solver);
+
+      std::vector<std::vector<double> >
+        y_res(ts.size(),
```
what does `res` mean?

syclik commented:
@wds15, there are some design questions I have. It still looks like this is much more complicated than it needs to be and I'm lost even trying to follow along.

`stan/math/rev/mat/functor/integrate_ode_cvodes.hpp` is the entry point from Stan's generated code. I think the function is the right abstraction and is the right design. Here's what the function signature is:
```{c++}
    template <typename F, typename T1, typename T2>
    std::vector<std::vector<typename stan::return_type<T1, T2>::type> >
    integrate_ode_cvodes(const F& f,
                         const std::vector<T1>& y0,
                         const double t0,
                         const std::vector<double>& ts,
                         const std::vector<T2> theta,
                         const std::vector<double>& x,
                         const std::vector<int>& x_int,
                         double rel_tol = 1e-10,
                         double abs_tol = 1e-10,
                         long int max_num_steps = 1e8,  // NOLINT(runtime/int)
                         const size_t solver = 1,
                         std::ostream* msgs = 0) {
```
Types: 
- `F` is a struct that has a functor for the ode. This is generated by stanc.
- `T1` can either be `stan::math::var` or `double`
- `T2` can either be `stan::math::var` or `double`

This is where I get lost. Here's rough high-level pseudocode for `integrate_ode_cvodes()`. I'm putting in comments where I'm confused.

1. check inputs. **no problems here**
2. convert `std::vector<T1> y0` to `std::vector<double> y0_dbl` and `std::vector<T2> theta` to `std::vector<double> t0_theta`. Add additional bookkeeping variables, `initial_var` and `param_var`, that hold type information whether `y0` or `theta` are vars. **I don't think this is necessary. We already know the types**
3. construct `stan::math::ode_model<F> ode` passing in
    - `f`: the ode functor
    - `theta_dbl`: the parameter values as double. **why double? why not the natural type?**
    - `x`: data
    - `x_int`: int data
   - `msgs`: output stream
   **I don't think this object is necessary and can be wrapped into the other object, `cvodes_integrator`**
4. construct `stan::math::cvodes_integrator<F> integrator` passing in:
    - `ode`: the `ode_model<F>` just constructed. **Can't we just pull this into the `cvodes_integrator` object?**
    - `y0_dbl`: the initial values as double. **I think this should just be `y0`**
    - `t0`: initial time
    - `initial_var::value`: an indicator that `y0` is a vector of `var`. **we shouldn't use this sort of construction. we can just pass in `y0` directly**
    - `param_var::value`: indicator that `theta` is a vector of `var`. **we shouldn't do this either**
    - `rel_tol`, `abs_tol`, `max_number_steps`: values controlling cvodes
    - `solver`: the integer that indicates the solver {0, 1, 2}
   **I think this should take care of building `ode` if we keep that object around**
5. create a `std::vector<std::vector<double>> y_res` with the size of `ts` and each of the inner `std::vector`s should have size `integrator.size()`
6. run `integrator.integrate_times(ts, y_res)`: this is a void function and fills `y_res`. **I think we should just combine this step and the next one and return the right thing**
7. return the result from `stan::math::decouple_ode_states(y_res, y0, theta)`. **I think `integrate_times` should just handle this and return the value. I don't think we need to pass in `y0` and `theta` -- these should already be known by the objects that were created. This would simplify the logic.**

Here's what I would suggest instead:
1. check inputs.
2. construct `cvodes_integrator` with everything for both `ode_model`, `cvodes_integrator`, and `decouple_states`
3. `return integrator.integrate_times(ts);`
Let me know if you think that won't work.


I have more inline comments coming.

syclik commented on stan/math/rev/mat/functor/cvodes_integrator.hpp:
```diff
@@ -0,0 +1,376 @@
+
+        // Instantiate CVode memory
+        if (solver == 0) {
+          cvode_mem_ = CVodeCreate(CV_ADAMS, CV_FUNCTIONAL);
```
For a future pull request, but I'm curious; is there a way to avoid calling `CVodeCreate` and `CVodeFree` again and again? If we made each solver its own class, could we possibly have some flyweight implementation?

bob-carpenter replied:
If the object created is expensive to create *and* has reusable
memory, it would make sense to have a singleton implementation.
Otherwise, there's really no point.   

Alternatively, we could enforce good C++ hygiene by using
a pointer-to-implementation (pimpl) behind a resource-allocation-is-init
(RAII) pattern, the same way that std::vector and Eigen::Matrix do.


> On Apr 20, 2016, at 11:17 AM, Daniel Lee <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/cvodes_integrator.hpp:
> 
> > +          S_((initial_var ? N_ : 0) + (param_var ? M_ : 0)),
> > +          size_(N_ * (S_+1)),
> > +          param_var_ind_(initial_var ? N_ : 0),
> > +          cvode_mem_(NULL),
> > +          state_(y0),
> > +          cvode_state_(N_VMake_Serial(N_,
> > +                                      &state_[0])),
> > +          cvode_state_sens_(NULL) {
> > +        stan::math::check_bounded("cvodes_integrator",
> > +                                  "solver", solver,
> > +                                  static_cast<size_t>(0),
> > +                                  static_cast<size_t>(2));
> > +
> > +        // Instantiate CVode memory
> > +        if (solver == 0) {
> > +          cvode_mem_ = CVodeCreate(CV_ADAMS, CV_FUNCTIONAL);
> 
> For a future pull request, but I'm curious; is there a way to avoid calling CVodeCreate and CVodeFree again and again? If we made each solver its own class, could we possibly have some flyweight implementation?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

wds15 replied:
I remember running some benchmarks and calling integrate_ode_cvodes between each time-points or running it in a single sweep - I did not see any difference. But cvode offers functionality to reuse a one initialized solver, it should be faster, but I could not measure it.

syclik commented on stan/math/rev/mat/functor/cvodes_integrator.hpp:
```diff
@@ -0,0 +1,376 @@
+          check_flag(CVDlsSetDenseJacFn(cvode_mem_, &ode::dense_jacobian),
+                      "CVDlsSetDenseJacFn");
+        }
+
```
can we move the `set_cvode_options()` call to here?

wds15 replied:
Nope, the order is really fixed here. This is how cvode dictates it to me.

syclik commented on stan/math/rev/mat/functor/ode_model.hpp:
```diff
@@ -0,0 +1,165 @@
+       */
+      inline
+      void operator()(const std::vector<double>& y,
+                      std::vector<double>& dy_dt,
```
We usually move the parameters to an argument that are used for output to the end of the list. Could you change the order here?

bob-carpenter replied:
I agree.  It's conventional to have the modified arguments
at the end if possible.

> On Apr 20, 2016, at 11:50 AM, Daniel Lee <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/ode_model.hpp:
> 
> > +          theta_(theta),
> > +          x_(x),
> > +          x_int_(x_int),
> > +          msgs_(msgs)
> > +      {}
> > +
> > +      /**
> > +       * Calculate the RHS of the ODE
> > +       *
> > +       * @param[in] y state of the ode system at time t.
> > +       * @param[out] dy_dt ODE RHS
> > +       * @param[in]  t time.
> > +       */
> > +      inline
> > +      void operator()(const std::vector<double>& y,
> > +                      std::vector<double>& dy_dt,
> 
> We usually move the parameters to an argument that are used for output to the end of the list. Could you change the order here?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

syclik commented on stan/math/rev/mat/functor/cvodes_integrator.hpp:
```diff
@@ -0,0 +1,376 @@
+       * @param[out] dy_dt ODE RHS at time t
+       * @param[in] t time
+       */
+      void rhs(const double y[], double dy_dt[], double t) const {
```
Change order of arguments

wds15 replied:
Ok, done. Although this was inspired from how the call to the coupled_ode_system looks like...

syclik commented on stan/math/rev/mat/functor/cvodes_integrator.hpp:
```diff
@@ -0,0 +1,376 @@
+      }
+
+      // Static wrapper for CVode callback
+      static int ode_rhs_sens(int Ns, realtype t,
```
Do these callback functions need to be `static`?

wds15 replied:
I am just doing the same as @betanalpha did, so I guess, yes these need to be static - Michael should know this he figured that bit out, not me.

betanalpha replied:
Yes.

On Apr 20, 2016, at 7:54 PM, Daniel Lee <notifications@github.com> wrote:

> In stan/math/rev/mat/functor/cvodes_integrator.hpp:
> 
> > +          ode_model_.jacobian_S(t, y_vec, fy, Jy);
> > +        }
> > +
> > +        if (initial_var_) {
> > +          for (size_t m = 0; m < N_; m++) {
> > +            // map NV_Vector to Eigen facilities
> > +            Eigen::Map<Eigen::VectorXd> yS_eig(NV_DATA_S(yS[m]), N_);
> > +            Eigen::Map<Eigen::VectorXd> ySdot_eig(NV_DATA_S(ySdot[m]), N_);
> > +
> > +            ySdot_eig = Jy * yS_eig;
> > +          }
> > +        }
> > +      }
> > +
> > +      // Static wrapper for CVode callback
> > +      static int ode_rhs_sens(int Ns, realtype t,
> Do these callback functions need to be static?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

syclik replied:
I think I see why this happened. This was to initialize it in the constructor? If it was done after the constructor, I'm guessing we don't need the static function?

> On Apr 20, 2016, at 6:08 PM, Michael Betancourt <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/cvodes_integrator.hpp:
> 
> > +          ode_model_.jacobian_S(t, y_vec, fy, Jy);
> > +        }
> > +
> > +        if (initial_var_) {
> > +          for (size_t m = 0; m < N_; m++) {
> > +            // map NV_Vector to Eigen facilities
> > +            Eigen::Map<Eigen::VectorXd> yS_eig(NV_DATA_S(yS[m]), N_);
> > +            Eigen::Map<Eigen::VectorXd> ySdot_eig(NV_DATA_S(ySdot[m]), N_);
> > +
> > +            ySdot_eig = Jy * yS_eig;
> > +          }
> > +        }
> > +      }
> > +
> > +      // Static wrapper for CVode callback
> > +      static int ode_rhs_sens(int Ns, realtype t,
> Yes.
> …
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

betanalpha replied:
No.  The static is needed to ensure that the method can be
accessed with a global function pointer, which then allows 
us to pass it to the C callbacks in CVODE(S).

On Apr 20, 2016, at 11:18 PM, Daniel Lee <notifications@github.com> wrote:

> In stan/math/rev/mat/functor/cvodes_integrator.hpp:
> 
> > +          ode_model_.jacobian_S(t, y_vec, fy, Jy);
> > +        }
> > +
> > +        if (initial_var_) {
> > +          for (size_t m = 0; m < N_; m++) {
> > +            // map NV_Vector to Eigen facilities
> > +            Eigen::Map<Eigen::VectorXd> yS_eig(NV_DATA_S(yS[m]), N_);
> > +            Eigen::Map<Eigen::VectorXd> ySdot_eig(NV_DATA_S(ySdot[m]), N_);
> > +
> > +            ySdot_eig = Jy * yS_eig;
> > +          }
> > +        }
> > +      }
> > +
> > +      // Static wrapper for CVode callback
> > +      static int ode_rhs_sens(int Ns, realtype t,
> I think I see why this happened. This was to initialize it in the constructor? If it was done after the constructor, I'm guessing we don't need the static function?
> …
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

bob-carpenter replied:
With static, you don't pass the class instance
in as an implicit first argument.  So presumably if it
works with it being static then it has to be static.

> On Apr 20, 2016, at 5:17 PM, wds15 <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/cvodes_integrator.hpp:
> 
> > +          ode_model_.jacobian_S(t, y_vec, fy, Jy);
> > +        }
> > +
> > +        if (initial_var_) {
> > +          for (size_t m = 0; m < N_; m++) {
> > +            // map NV_Vector to Eigen facilities
> > +            Eigen::Map<Eigen::VectorXd> yS_eig(NV_DATA_S(yS[m]), N_);
> > +            Eigen::Map<Eigen::VectorXd> ySdot_eig(NV_DATA_S(ySdot[m]), N_);
> > +
> > +            ySdot_eig = Jy * yS_eig;
> > +          }
> > +        }
> > +      }
> > +
> > +      // Static wrapper for CVode callback
> > +      static int ode_rhs_sens(int Ns, realtype t,
> 
> I am just doing the same as @betanalpha did, so I guess, yes these need to be static - Michael should know this he figured that bit out, not me.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

bob-carpenter replied:
Sorry --- hadn't read far enough.  Yes, that's what I meant.
The global means it doesn't need the class instance.

> On Apr 20, 2016, at 6:20 PM, Michael Betancourt <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/cvodes_integrator.hpp:
> 
> > +          ode_model_.jacobian_S(t, y_vec, fy, Jy);
> > +        }
> > +
> > +        if (initial_var_) {
> > +          for (size_t m = 0; m < N_; m++) {
> > +            // map NV_Vector to Eigen facilities
> > +            Eigen::Map<Eigen::VectorXd> yS_eig(NV_DATA_S(yS[m]), N_);
> > +            Eigen::Map<Eigen::VectorXd> ySdot_eig(NV_DATA_S(ySdot[m]), N_);
> > +
> > +            ySdot_eig = Jy * yS_eig;
> > +          }
> > +        }
> > +      }
> > +
> > +      // Static wrapper for CVode callback
> > +      static int ode_rhs_sens(int Ns, realtype t,
> 
> No. The static is needed to ensure that the method can be accessed with a global function pointer, which then allows us to pass it to the C callbacks in CVODE(S).
> …
> On Apr 20, 2016, at 11:18 PM, Daniel Lee ***@***.***> wrote: In stan/math/rev/mat/functor/cvodes_integrator.hpp: > + ode_model_.jacobian_S(t, y_vec, fy, Jy); > + } > + > + if (initial_var_) { > + for (size_t m = 0; m < N_; m++) { > + // map NV_Vector to Eigen facilities > + Eigen::Map<Eigen::VectorXd> yS_eig(NV_DATA_S(yS[m]), N_); > + Eigen::Map<Eigen::VectorXd> ySdot_eig(NV_DATA_S(ySdot[m]), N_); > + > + ySdot_eig = Jy * yS_eig; > + } > + } > + } > + > + // Static wrapper for CVode callback > + static int ode_rhs_sens(int Ns, realtype t, I think I see why this happened. This was to initialize it in the constructor? If it was done after the constructor, I'm guessing we don't need the static function? … — You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

syclik commented:
@wds15, after thinking about this a bit, can we simplify the code by:

1. going from one `integrate_ode_cvodes()` call to one per solver, so 3 functions
2. simplify `ode_model` and `cvodes_integrator` by
    - merging both together. having two things doesn't actually separate out functionality, so I'm not sure what this does.
    - make `cvodes_integrator` by simplifying it by using types. Right now, there are just a lot of conditionals based on types. I'd rather have the specializations just do the right thing. There might be some more duplicated code, but following the logic would be much, much clearer.

syclik commented:
My current suggestion is to go forward with v2.10.0 without this pull request. @betanalpha's implementation using a coupled system will be exposed. We should address how the functions are exposed through the language and do that before the release of stan-dev/stan.

To fix this pull request, I think it'll take a bit of time. I'd rather not hold up everything for this pull request. Especially since I've got other things that are on my schedule that need to be done soon. @wds15, let me know if you need a clearer explanation of how to tease apart the implementation so it's easier to follow.

If you think this should be the highest priority and we can't release 2.10 without it, please chime in. I'm happy to redistribute my priorities if it's significant.

wds15 commented:
Hi Daniel!

I did try hard to address your points and changed the logic accordingly, i.e. the natural types are now given to the cvodes_integrator and the integrator returns the solution with the correct type directly.

Moreover, I do think that the ode_model object is a very powerful way of abstracting things for many reasons:

- it allows us to refactor the odeint integrator, going from ~800 lines of code to around ~100. Please have a look at the branch feature/issue-288-refactor-odeint which is a demonstration of this refactoring

- we can easily insert into Stan analytical Jacobian information by partially specialising this object. This will be important for very stiff systems where autodiff will quickly become a bottleneck.

All other requests from your end - I think I did address them.

To simplify the adoption for now, I would suggest to rename integrate_ode_cvodes to integrate_ode_bdf and I remove the solver option from the cvodes_integrator - we can add that later again.

Should this hold up 2.10? The answer depends. If you ask anyone doing stiff ODEs (and I do), then yes we should get this in now. My tests have shown that a 2-cmt oral dosing model (very simple) will freeze the current cvode integrator. I am afraid that Stan wont be usable for any stiff application when we leave it in the current state. So yes, to me this pull is high priority and I can arrange some time to make this happen now.

syclik commented:
there's nothing on the branch for 288 yet. I'm curious to see what it looks like.

wds15 commented:
Oh, sorry... too late for me... now you got it. And yes, maybe this code example helps to explain the logic. Without the CVODES stuff in the way things are very straightforward to write down usig Eigen etc.

syclik commented:
I took a look at the branch and don't see how it's more useful.

Here's what I think needs to happen.

1. Split `integrate_ode_cvodes()` into 3 calls. This will be the entry point from the Stan language. I think we decided on calling it the algorithm name, so something like:
    - `integrate_ode_adams_moulton()`
    - `integrate_ode_bdf()`
    - `integrate_ode_bdf_stability()` -- or whatever the right way to describe it is.
   Inside each of these functions, skip the construction of `ode_model`.
2. change the constructor of `cvodes_integrator` to accept the additional arguments of `ode_model`. The top level functions don't need the `ode_model` at all. Also change the types to accept `y0` and `theta` directly instead of double versions.
3. if you really want an `ode_model`, I think it should be changed to be used as what the static callbacks for CVODES calls. So, it should have these functions:
    - `rhs`
    - `rhs_sens`
    - `dense_jacobian`
    I think that should be passed in as the callback to make the `cvodes_integrator` class have one purpose: set up CVODES and run the integrator. Right now, it mixes the setup and running and the running actually gets split logically across `cvodes_integrator` and `ode_model`.
    So, my suggestion is to separate out the concerns and have the classes line up with the separation of work.
4. For `ode_model`, I would also template it and implement it more simply instead of trying to be clever with the conditionals based on type within the functions.

I hope that makes sense. I won't be able to help out for a couple weeks on this. I agree with the functionality, but it was really, really hard to follow the code because the scope of what the classes dealt  didn't line up with the separation of work.

I think we move forward with v2.10 without this and this will be available as soon as it gets done.

wds15 commented:
Hi Daniel!

I think I covered all your points:

1. I renamed from integrate_ode_cvodes to integrate_ode_bdf. I don't expose the other integrators right now for the reason that tests need to be written. All tests have been brought in line with with _bdf integrator function. 

2. The cvodes_integrator constructor now takes care of everything, i.e. the ode_model is created inside cvodes_integrator. Moreover, I pass in now the ts vector as well. The ts vector can anyway not be changed as t0 is fixed. The types of y0 and theta are the natural types in cvodes_integrator now.

3. I really would like to have ode_model, yes. See below.

4. I am not sure how ode_model can be more simple. What I did is that you can pass in the natural type of theta, but internally this is converted to a double, because this is what is needed.

I thought to have separated how the work is done, i.e. cvodes_integrator object deals with setting up CVODES and it feeds into CVODES whatever is needed using Stan functionality.

The sole purpose of ode_model is to make the mathematical objects needed accessible in a flexible and convenient way. So if you could comment on the following, then I can understand you concerns hopefully better:

- The objects defined in this model allow for a far more convenient writing of the code, i.e. the odeint stuff goes down from ~800 lines of code to ~100.

- Moreover, it brings the huge benefit that the code gets much easier to read (to me), since instead of endless for-loops I can use matrix products which are much more straightforward and less error-prone.

- The ode_model object will also allow us to elegantly enable an important feature if we want to make Stan a high-performance tool for stiff ODE modeling. That is, analytic Jacobians can easily be inserted into the code with the ode_model object. This is very important for stiff systems which require a frequent evaluation of the Jacobian and hence autodiff becomes a bottleneck. I have run some benchmarks and have seen speedups of 2-3x by doing this.

As I mentioned, I will post on stan-dev a bit of LaTeX which will explain the need for ode_model. 

Ultimately, if you deem the ode_model pre-mature, but would say that we should defer the discussion of having ode_model, then I could quickly move those functions as private member functions into cvodes_integrator to make a quick adoption possible at this stage. Although I think this would not be the right way to go forward.

wds15 commented:
Not sure why doxygen flags a warning in cvodes_integrator being wrongly documented. It refers to ode_model being given wrongly to the constructor, but this not what is done, so I am not sure if Jenkins pulled out the right version of the file or I misunderstand.

syclik commented on stan/math/rev/mat/functor/cvodes_integrator.hpp:
```diff
@@ -0,0 +1,383 @@
+       * While the first N states correspond to y, all the remaining
+       * are optional and depend on flags given to the constructor.
+       *
+       * @param[in] ode_model functor.
```
Doxygen is properly flagging this as a problem. There's no `ode_model` argument to the `cvodes_constructor`.
`ode_model` needs to be replaced with `f`. Every `@param` should line up with every parameter. 

This is just nitpicking and you don't need to do this. Since this is a constructor and all things are inputs, I think we should just remove all the `[in]`. I really hope none of our code is modifying inputs through constructors....

syclik commented:
Sebastian, thanks for making these changes!

So, I now understand a lot of what's going on. I think we should still release v2.10 without this because it'll take a little more work and we shouldn't just hold up releases for features that aren't quite ready. 

Here is what I think we should do for this to be ready:

`ode_model`:
1. Change the name of `ode_model` to `ode`. Unless you have a very good reason for naming it model. I don't know what the model portion of it is and I think just calling it `ode` is as descriptive as `ode_model`.
2. In the constructor, remove the template parameter, `T`. Change the type of `theta` from `std::vector<T>` to `std::vector<double>`.
3. Change the function signature for `jacobian_S` to
```
template <typename Derived>
void jacobian_S(const double t, const std::vector<double>& y,
                          Eigen::VectorXd& dy_dt,
                          Eigen::MatrixBase<Derived>& Jy) const
```
I've looked through all the uses and I don't see where `dy_dt` can not be `Eigen::VectorXd`. The added generality made it harder to understand -- I didn't know if it could actually be a matrix or some other shape and what that meant.
4.  Change the function signature for `jacobian_SP` to
```
template <typename Derived>
void jacobian_SP(const double t, const std::vector<double>& y,
                             Eigen::VectorXd& dy_dt,
                             Eigen::MatrixBase<Derived>& Jy,
                             Eigen::MatrixBase<Derived>& Jtheta) const
```
The only use of `jacobian_SP` uses `Eigen::MatrixXd`, but I see where you're going with this and I think having `MatrixBase<Derived>` is ok.
5. Is there any reason `ode_model` isn't using `double[]` directly instead of `std::vector`? From `cvodes_integrator`, it looks like we're shuffling back and forth when it seems like the right place to shuffle this is `ode_model`. `ode_model` seems like its function is to be an abstraction providing `operator()`, `jacobian_S()`, and `jacobian_SP()` from the ode functor `f`.
Overall, I see where you've gone with the `ode_model` and think it's good. I think it's building too much generality.


`cvodes_integrator`:
I think this is where it's trying to do too much. Here's what I see that it's doing. Sebastian, please feel free to correct me where I'm incorrect:
1. configure 3 different CVODES integrators. We are only exposing one for this pull request.
2. handle 4 different inputs which have different control flows. This is for the type of (inital value, parameter value): (double, double), (double, var), (var, double), (var, var)

I'm ok with 1), but I think we can make the code simpler for 2). Here are concrete steps I think we can take:

1. create 3 functions for creating `cvode_mem_`. For example, something like: `cvode_create_adams`, `cvode_create_bdf`, `cvode_create_bdf_with_stald`. Change the constructor so it calls those functions with simple if statements.
2. Let's replace `initial_var` with `is_var<T1>::value` and `param_var::value` with `is_var<T2>::value`. You'll need to include `stan/math/rev/scal/meta/is_var.hpp`. (sorry -- we don't have a comprehensive list of our template meta programs, but `is_var` exists.)
3. Can we call `T1` something like `T_initial` and `T2` something like `T_param`?
4. I think I finally got my head around the complexity. It looks like we do different logic for when `S_ == 0` and when `S_ > 0`. (`S_ > 0` when either `T1` or `T2` are vars.) I realize that with the templating, we can't use object oriented programming to use a base class then build on it. I'm really wanting to split out the 4 cases. It would make logic a lot simpler. I understand what's going on, but it's making me dizzy trying to follow along. I'm looking at `rhs_sens` and instead of there just being 4 blocks, it's multiple if statements. Can we simplify with a little duplication, but just breaking it into those 4 blocks?

Does that make sense? Sebastian, please feel free to question what I've suggested. I might not understand the use case beyond what's here in front of us. I'd suggest template specializations, but I think that defeats the purpose of using templates here. I want to do things using composition of objects, but I don't think there's a good way to do it without the curiously recurring template pattern (CRTP), which is a bit overkill.

wds15 commented:
Hi Daniel!

I am very glad to read these lines as the code starts to make sense to you. Great. Your comments make sense and I will need to work through them carefully. Here are my thoughts for it:

General remark:
I introduced the templated MatrixBase stuff following what the Eigen documentation suggests for functions dealing with Eigen objects. This gives you the greatest flexibility in that you can pass into it Map stuff and real data. The Map functionality is just so amazing as it can wrap around any chunk of memory.

And yes, in Eigen everything is a Matrix - including vectors.

ode_model:
1. I added "model" to indicate that this is a bit more than just the ode RHS, i.e. it is the ODE RHS with Jacobians. Maybe ode_system? I can live with ode... but that somehow would miss the fact that this is more than just the ODE RHS and could me misleading eventually. So let me know what would be clear to you given this aspect.

2. Ok I take T out of the constructor.

3. If we want to use the ode object inside the RK45 or other integrators, then we should let dy_dt also be a templated argument of the function as I did originally, this avoids copying things around as I can use Map's. Moreover, I can wrap around the dy_dt[] C arrays from CVODES with a much nicer C++ construct from Eigen, so I avoid arrays here.

4. If you only template for Derived, then you force Jy and Jtheta to be of the same type. That is we can not anymore pass in Jy as Map and Jtheta as Matrix. I think my code does not need this flexibility, but I did not see a reason to not give this flexibility. So my question to you: What rational you have to remove the extra template argument - keep it as simple as possible? Just for my understanding.

5. I think of ode_model as a nice and clean C++ object. Something which can be reused in other parts of Stan, so I don't want to see those C arrays there; instead I much prefer to use templated MatrixBase things and pass in Eigen Map's which is much cleaner (and BTW, this is the reason why I made dy_dt also a Eigen MatrixBase thing as this avoids copying stuff around from the raw CVODES memory).

My intent is to keep all the C array handling stuff inside cvodes_integrator and nowhere else. I see C arrays as something to avoid in C++ code and hence would keep this local to where it has to be done like that.

cvodes_integrator:
1. Can I keep the set_cvode_options which would be called from the three different cvode_create_* functions?

2. I did not know of is_var - of course, that makes a lot of sense.

3. OK, I rename

4. First, there are actually only 3 cases to distinguish in rhs_sens; for the case of S=0 (no sensitivities) the rhs_sens is never called at all as the sensitivity system isn't even initialized. I tried to avoid code duplication, yes; but we can split, sure.

In fact I was also thinking hard about a common base class and my thoughts were also going in the direction of CRTP, but I could not make sense of it. How about the following compromise:

The static ode_rhs_sens call generates a call to

`rhs_sens(...old arguments..., initial_var(), param_var());
`
Then we have three different rhs_sens functions with the signature

```
rhs_sens(...old arguments..., true_type, true_type); => initials and params vary
rhs_sens(...old arguments..., true_type, false_type); => only initials vary
rhs_sens(...old arguments..., false_type, true_type); => only parmas vary

```
The individual functions could be written such that they share some code to avoid code duplication.

Does that make sense?

Wrt to 2.10: If we continue at this pace, then we are in good shape, I think. If you have already taken the decision, fine with me; another option would be that I work these things out these days and on Tuesday we decide with the team what to do. Just a thought.

To me it is much clearer what you intent. To be honest, it is sad to see that I was first blocked for ~3 weeks to do benchmarking+precision tests. This time was time lost for this important design discussion.

bob-carpenter commented:
> On Apr 22, 2016, at 3:16 AM, wds15 <notifications@github.com> wrote:
> 
> Hi Daniel!
> 
> I am very glad to read these lines as the code starts to make sense to you. Great. Your comments make sense and I will need to work through them carefully. Here are my thoughts for it:
> 
> General remark:
> I introduced the templated MatrixBase stuff following what the Eigen documentation suggests for functions dealing with Eigen objects. This gives you the greatest flexibility in that you can pass into it Map stuff and real data. The Map functionality is just so amazing as it can wrap around any chunk of memory.
> 
> And yes, in Eigen everything is a Matrix - including vectors.

Agreed that Map<> is great.

Not quite.  The block() operations and a few other edge cases
won't match and will require explicit evaluate steps.  We've discussed
this before on the list, and I think it's a serious problem for
writing Eigen functions.


> ode_model:
> 1. I added "model" to indicate that this is a bit more than just the ode RHS, i.e. it is the ODE RHS with Jacobians. Maybe ode_system? I can live with ode... but that somehow would miss the fact that this is more than just the ODE RHS and could me misleading eventually. So let me know what would be clear to you given this aspect.
> 
> 	• Ok I take T out of the constructor.
> 
> 	• If we want to use the ode object inside the RK45 or other integrators, then we should let dy_dt also be a templated argument of the function as I did originally, this avoids copying things around as I can use Map's. Moreover, I can wrap around the dy_dt[] C arrays from CVODES with a much nicer C++ construct from Eigen, so I avoid arrays here.
> 
> 	• If you only template for Derived, then you force Jy and Jtheta to be of the same type. That is we can not anymore pass in Jy as Map and Jtheta as Matrix. I think my code does not need this flexibility, but I did not see a reason to not give this flexibility. So my question to you: What rational you have to remove the extra template argument - keep it as simple as possible? Just for my understanding.
> 

I think it's fine to generalize inputs.  You just have
to be careful to test all the permutations and make sure
that you have the right intermediate promote_args types for
intermediate assignments.

> 	• I think of ode_model as a nice and clean C++ object. Something which can be reused in other parts of Stan, so I don't want to see those C arrays there; instead I much prefer to use templated MatrixBase things and pass in Eigen Map's which is much cleaner (and BTW, this is the reason why I made dy_dt also a Eigen MatrixBase thing as this avoids copying stuff around from the raw CVODES memory).

I'd also prefer real C++ types where possible.  

With Eigen::Map, you still have to manage your own memory,
so it's not neatly encapsulated like a real Eigen::Matrix
or std::vector (not saying not to use it, just that you still
have to be careful).

> My intent is to keep all the C array handling stuff inside cvodes_integrator and nowhere else. I see C arrays as something to avoid in C++ code and hence would keep this local to where it has to be done like that.

+1.

> cvodes_integrator:
> 1. Can I keep the set_cvode_options which would be called from the three different cvode_create_* functions?
> 
> 	• I did not know of is_var - of course, that makes a lot of sense.
> 
> 	• OK, I rename
> 
> 	• First, there are actually only 3 cases to distinguish in rhs_sens; for the case of S=0 (no sensitivities) the rhs_sens is never called at all as the sensitivity system isn't even initialized. I tried to avoid code duplication, yes; but we can split, sure.
> 
> In fact I was also thinking hard about a common base class and my thoughts were also going in the direction of CRTP, but I could not make sense of it.

I can help with that if you have a use case.  The Josuttis
and Vandevoorde books is great for all these C++ tempalte
patterns.

> How about the following compromise:
> 
> The static ode_rhs_sens call generates a call to
> 
> rhs_sens(...old arguments..., initial_var(), param_var());
> 
> Then we have three different rhs_sens functions with the signature
> 
> rhs_sens(...old arguments..., true_type, true_type); => initials and params vary
> rhs_sens(...old arguments..., true_type, false_type); => only initials vary
> rhs_sens(...old arguments..., false_type, true_type); => only parmas vary


What's true_type and false_type?

Why not just write three functions rather than overloading one?

> The individual functions could be written such that they share some code to avoid code duplication.
> 
> Does that make sense?
> 
> Wrt to 2.10: If we continue at this pace, then we are in good shape, I think. If you have already taken the decision, fine with me; another option would be that I work these things out these days and on Tuesday we decide with the team what to do. Just a thought.

We can release as soon as we have features ready.

The only thing that holds up our releases is that we are
inevitably waiting for just one more thing every time we
release. 

> To me it is much clearer what you intent. To be honest, it is sad to see that I was first blocked for ~3 weeks to do benchmarking+precision tests. This time was time lost for this important design discussion.

Those tests are important!

- Bob

wds15 commented:
The current code has some tests which break as the interface should settle first, I think.

I tried to implement CRTP, but aborted this mission as too many things would need to be defined in multiple places (maybe I missed something though). Instead I coded what I suggested, i.e. via function overloading the right case is selected and we get a clear-cut type-to-code mapping, which I guess is what you want.

I did have to stick with is_same from boost as it has the nifty additional feature that it not only provides runtime information, but also compile time information via the true_type and false_type structs. These are always cheap to instantiate which I don't know about the var instance.

Finally, I just posted on stan-dev my rationale for introducing the ode_model object (or however we want to call it). I hope this makes it easier to follow the logic, and others can give their input which won't hurt I guess.

I just saw to forget to rename the template parameter names... will do that next time.

bob-carpenter commented:
> On Apr 25, 2016, at 5:09 PM, wds15 <notifications@github.com> wrote:
> 
> The current code has some tests which break as the interface should settle first, I think.
> 
> I tried to implement CRTP, but aborted this mission as too many things would need to be defined in multiple places (maybe I missed something though).

CRTP doesn't require code duplication.  Here's the pattern:

https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern

...

>  we get a clear-cut type-to-code mapping, which I guess is what you want.

It's what the compiler demands.

> I did have to stick with is_same from boost as it has the nifty additional feature that it not only provides runtime information, but also compile time information via the true_type and false_type structs. These are always cheap to instantiate which I don't know about the var instance.

Boost's is_same is purely compile time:

http://www.boost.org/doc/libs/1_60_0/libs/type_traits/doc/html/boost_typetraits/reference/is_same.html

- Bob

wds15 commented:
Hi Bob,

I definetly tried CRTP... Can I send you the source which I stopped developing and you comment in the cpp file how to do it better? My main problem was that in the base class it gets vers tedious to access members of the derived, i.e. I would have ended up written tons of static cast's, but if there is an easy way which i overlook, then i am happy to rewrite once more.

The solution now works with overloading which can be resolved at compile time such that no performance hit is taken at all.

Thanks,
Sebastian

wds15 commented:
So here it goes. Now the cvodes_integrator uses the natural types which have been passed in for the initials and the parameters. The y0 and theta variables are then used as dummy arguments (they are not used in the rhs_sens function itself) to rhs_sens in order to go to the right implementation of the sensitivity rhs.

Moreover, I renamed the ode_model to ode_system and renamed T1/T2 to T_initial and T_param throughout.

I hope this is what we all called being on the same page yesterday...

Tests do break at the moment as the interface for cvodes_integrator has been changing. BTW, is there a way to declare the callbacks for CVODES (the static functions) as private? If this is possible (not sure), then I would suggest to do so in order to minimize what gets exposed to the outside.

Feedback would be very welcome. Thanks.

wds15 commented:
Code is now fully tested, i.e. I added tests for decouple_ode_states and the ode_system. The decouple_ode_states tests were easily derived from similar code which was formely used to test the coupled_ode_system code which brought its own decoupling operation. The tests for the ode_system are new and as these are the first tests I ever wrote from scratch for stan-math, some check would be great.

Moreover, I simplified the possible types somewhat for the jacobian function of ode_system. This was recommended by @syclik and reduces the combinatorial burden to test different combinations of types going into the function. In case we need the extra flexibility we may add it later along with the tests.
